lib/puppetserver/ca/create_action.rb

```diff
@@ -0,0 +1,253 @@
+      BANNER = <<-BANNER
+Usage:
+  puppetserver ca create [--help]
+  puppetserver ca create [--config] --certname CERTNAME[,ADDLCERTNAME]
+
+Description:
+Creates a new certificate signed by the intermediate CA
+and stores generated keys and certs on disk.
+
+To determine the target location, the default puppet.conf
+is consulted for custom values. If using a custom puppet.conf
+provide it with the --config flag
+Options:
+BANNER
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def run(input)
+        certnames = input['certnames']
+
+        # Validate config_path provided
+        config_path = input['config']
+        if config_path
+          errors = Puppetserver::Utils::FileUtilities.validate_file_paths(config_path)
+          return 1 if log_possible_errors(errors)
+        end
+
+        # Load, resolve, and validate puppet config settings
+        puppet = PuppetConfig.parse(config_path)
+        return 1 if log_possible_errors(puppet.errors)
+
+        # Load most secure signing digest we can for csr signing.
+        signer = Puppetserver::Utils::SigningDigest.new
+        return 1 if log_possible_errors(signer.errors)
+
+        # Make sure we have all the directories where we will be writing files
+        Puppetserver::Utils::FileUtilities.ensure_dir(puppet.settings[:certdir])
+        Puppetserver::Utils::FileUtilities.ensure_dir(puppet.settings[:privatekeydir])
+        Puppetserver::Utils::FileUtilities.ensure_dir(puppet.settings[:publickeydir])
+
+        passed = generate_certs(certnames, puppet.settings, signer.digest)
+        return passed ? 0 : 1
+      end
+
+      def log_possible_errors(maybe_errors)
+        errors = Array(maybe_errors).compact
+        unless errors.empty?
+          @logger.err "Error:"
+          errors.each do |message|
+            @logger.err "    #{message}"
+          end
+          return true
+        end
+      end
+
+
+      # Create crls and keys, then submit them to CA, request for the CA to sign
+      # them, and finally download the signed certificates from the CA.
+      def generate_certs(certnames, settings, digest)
+        name_csr_hash = generate_name_csr_hash(certnames, settings, digest)
+
+        submit_certificate_requests(name_csr_hash, settings)
+        sign_certs(certnames, settings)
+        download_certificates_from_ca(certnames, settings)
+      end
+
+      # Generate a csr and 2 keys for each certname and return a hash in format
+      # of {certname1 => csr1, certname2 => crs2}
+      def generate_name_csr_hash(certnames, settings, digest)
+        certname_csr_hash = {}
+        host = Puppetserver::Ca::Host.new(digest)
+        certnames.each do |certname|
+          new_private_key = host.create_private_key(settings[:keylength])
+          new_public_key = new_private_key.public_key
+          save(new_private_key, certname, settings[:privatekeydir])
+          save(new_public_key, certname, settings[:publickeydir])
```

Comment:
So, I shouldn't really save the keys until I'm sure we will get certificates made that correspond to those keys.  HOwever, I don't like the idea of holding on to those key strings until all certs have been returned.  Should I?  I guess I could save them in a hash with certname keys and then return all the certnames for the certs generated and then do it in a loop?  Better ideas?